src/core/credGrainView.js

```diff
@@ -54,77 +54,137 @@ export type ParticipantsOptions = {|
+          ),
+        };
+      })
     );
   }
 
+  _calculateGrainEarnedPerInterval(account: Account): $ReadOnlyArray<?Grain> {
+    return this._intervals.map((interval) => {
+      let grain = G.ZERO;
+      account.allocationHistory.forEach((allocationReceipt) => {
+        if (
+          interval.startTimeMs < allocationReceipt.credTimestampMs &&
+          allocationReceipt.credTimestampMs <= interval.endTimeMs
+        )
+          grain = G.add(grain, allocationReceipt.grainReceipt.amount);
+      });
+      return grain;
+    });
+  }
+
   withTimeScope(
-    startTimeMs: number,
-    endTimeMs: number
+    startTimeMs: TimestampMs,
+    endTimeMs: TimestampMs
   ): TimeScopedCredGrainView {
     return new TimeScopedCredGrainView(
-      this._credGraph,
-      this._ledger,
+      this._participants,
+      this._intervals,
       startTimeMs,
       endTimeMs
     );
   }
 
   intervals(): IntervalSequence {
-    return this._defaultTimeScope.intervals();
+    return this._intervals;
   }
 
-  participant(id: IdentityId): ParticipantCredGrain | null {
-    return this._defaultTimeScope.participant(id);
-  }
-
-  participants(
-    options: ParticipantsOptions
-  ): $ReadOnlyArray<ParticipantCredGrain> {
-    return this._defaultTimeScope.participants(options);
+  participants(): $ReadOnlyArray<ParticipantCredGrain> {
+    return this._participants;
   }
 }
 
+/**
+ * This class's contructor stores a continuous subset of the originalIntervals
+ * and participant cred/grain data where intervals are only included if their
+ * start and end times are both within the provided startTimeMs and endTimeMs,
+ * inclusively.
+ */
 export class TimeScopedCredGrainView {
-  _credGraph: CredGraph;
-  _ledger: Ledger;
-  _startTimeMs: TimestampMs;
-  _endTimeMs: TimestampMs;
+  _participants: $ReadOnlyArray<ParticipantCredGrain>;
+  _intervals: IntervalSequence;
 
   constructor(
-    credGraph: CredGraph,
-    ledger: Ledger,
+    originalParticipants: $ReadOnlyArray<ParticipantCredGrain>,
+    originalIntervals: IntervalSequence,
     startTimeMs: TimestampMs,
     endTimeMs: TimestampMs
   ) {
-    this._credGraph = credGraph;
-    this._ledger = ledger;
-    this._startTimeMs = startTimeMs;
-    this._endTimeMs = endTimeMs;
-  }
+    let inclusiveStartIndex = originalIntervals.findIndex(
+      (interval) => startTimeMs <= interval.startTimeMs
+    );
+    if (inclusiveStartIndex === -1)
+      inclusiveStartIndex = originalIntervals.length;
 
-  participants(
-    options: ParticipantsOptions
-  ): $ReadOnlyArray<ParticipantCredGrain> {
-    const _ = options;
-    throw new Error("not yet implemented");
-  }
+    const exclusiveEndIndex =
+      findLastIndex(
+        originalIntervals,
+        (interval) => interval.endTimeMs <= endTimeMs
+      ) + 1;
 
-  participant(id: IdentityId): ParticipantCredGrain | null {
-    const _ = id;
-    throw new Error("not yet implemented");
+    this._intervals = deepFreeze(
+      intervalSequence(
```

Comment:
Nit: we already froze the original interval sequence in the outer constructor, so we could just do `Object.freeze` and get the same level of safety here. Likely too clever since it might confuse future readers, and the performance implications are negligible. :)